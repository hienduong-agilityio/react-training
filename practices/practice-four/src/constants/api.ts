export const HTTP_METHODS = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE'
};

export const ROOT_API = import.meta.env.VITE_API_ENDPOINT;

export const API_URL = {
  PROJECT: `${ROOT_API}/projects`
};
