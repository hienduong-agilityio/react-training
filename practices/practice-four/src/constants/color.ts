// Consider using enums instead of constants
export const COLORS_CLASS = {
  DEFAULT: 'color-default',
  PRIMARY: 'color-primary',
  SUCCESS: 'color-success',
  WARNING: 'color-warning',
  DANGER: 'color-danger '
};
