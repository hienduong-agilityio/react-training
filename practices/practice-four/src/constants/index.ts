// Index constants files
export * from './api';
export * from './color';
export * from './filerSearchOption';
export * from './route';
export * from './rowPerPage';
export * from './store';
export * from './validationMessages';
