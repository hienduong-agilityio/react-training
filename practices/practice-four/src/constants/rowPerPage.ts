export const ROWS_PER_PAGE_OPTIONS = [
  { value: 5, label: '5' },
  { value: 10, label: '10' },
  { value: 15, label: '15' },
  { value: 20, label: '20' },
  { value: 30, label: '30' },
  { value: 40, label: '40' },
  { value: 50, label: '50' }
];
