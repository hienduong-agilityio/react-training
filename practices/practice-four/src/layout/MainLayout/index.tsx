// Libraries
import { Outlet } from 'react-router-dom';

// Layouts
import { Sidebar, Navbar, TabLayout } from '@/layout';

/**
 * MainLayout component
 *
 * @returns {JSX.Element} The main layout structure with navigation components and content area.
 */
export const MainLayout = (): JSX.Element => {
  return (
    <main className='flex'>
      <Sidebar />
      <div className='flex flex-col w-full'>
        <Navbar />
        <TabLayout />
        <Outlet />
      </div>
    </main>
  );
};
