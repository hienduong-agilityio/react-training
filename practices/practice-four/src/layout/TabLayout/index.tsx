// Libraries
import { generateBreadcrumbItems } from '@/helpers/breadcrumb';
import { useMemo } from 'react';
import { useLocation, useParams } from 'react-router-dom';

// Components
import { Breadcrumbs } from '@/components';

/**
 * TabLayout component
 *
 * @returns {JSX.Element} - The TabLayout element
 */
export const TabLayout = (): JSX.Element => {
  const location = useLocation();
  const { id: projectId } = useParams();

  const breadcrumbItems = useMemo(
    () => generateBreadcrumbItems(location.pathname, projectId),
    [location.pathname, projectId]
  );

  return (
    <div className='px-4 py-2'>
      <Breadcrumbs items={breadcrumbItems} />
    </div>
  );
};
