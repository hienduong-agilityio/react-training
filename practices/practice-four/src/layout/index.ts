// Index Layout files
export * from './MainLayout';
export * from './NavBar';
export * from './ProjectLayout';
export * from './SideBar';
export * from './TabLayout';
