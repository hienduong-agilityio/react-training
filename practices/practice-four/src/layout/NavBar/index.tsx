// Libraries
import { Link } from 'react-router-dom';

// Components
import { Avatar } from '@/components';

// Svg
import notifyIcon from '@public/images/notificationIcon.svg';
import helpIcon from '@public/images/helpIcon.svg';

// Constants
import { ROUTE } from '@/constants';

/**
 * Navbar components
 *
 * @returns {JSX.Element} The Navbar component.
 */
export const Navbar = (): JSX.Element => {
  return (
    <nav className='bg-light w-full border'>
      <div className='flex flex-wrap items-center justify-between p-4'>
        <Link to={ROUTE.PROJECT} className='text-2xl font-semibold text-gray-900'>
          Projects
        </Link>
        <div className='flex items-center gap-6'>
          <img src={notifyIcon} className='cursor-not-allowed' alt='Notification' />
          <img src={helpIcon} className='cursor-not-allowed' alt='Help' />
          <Avatar customClass='rounded-full cursor-not-allowed' name='HienDuong' />
        </div>
      </div>
    </nav>
  );
};
