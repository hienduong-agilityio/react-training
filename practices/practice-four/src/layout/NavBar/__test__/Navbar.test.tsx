// Libraries
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

// Components
import { Navbar } from '@/layout';

// Constants
import { ROUTE } from '@/constants';

// Mock the Avatar component
jest.mock('@/components/common/Avatar', () => ({ customClass, name }: { customClass: string; name: string }) => (
  <div className={customClass}>Mocked Avatar - {name}</div>
));

// Mock the images
jest.mock('@public/images/notificationIcon.svg', () => 'mock-notify-icon');
jest.mock('@public/images/helpIcon.svg', () => 'mock-help-icon');

/**
 Navbar
  √ should render the Projects link
  √ should render the NavBar children components
*/
describe('Navbar', () => {
  const renderNavbar = () =>
    render(
      <MemoryRouter>
        <Navbar />
      </MemoryRouter>
    );

  it('should match the snapshot', () => {
    const { asFragment } = renderNavbar();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the NavBar children components', async () => {
    renderNavbar();
    const projectLink = await screen.findByText('Projects');

    expect(projectLink).toBeInTheDocument();
    expect(projectLink).toHaveAttribute('href', ROUTE.PROJECT);

    const notificationIcon = screen.getByAltText('Notification');
    expect(notificationIcon).toBeInTheDocument();
    expect(notificationIcon).toHaveAttribute('src', 'mock-notify-icon');

    const helpIcon = screen.getByAltText('Help');
    expect(helpIcon).toBeInTheDocument();
    expect(helpIcon).toHaveAttribute('src', 'mock-help-icon');

    const avatar = screen.getByText('Mocked Avatar - HienDuong');
    expect(avatar).toBeInTheDocument();
    expect(avatar).toHaveClass('rounded-full cursor-not-allowed');
  });
});
