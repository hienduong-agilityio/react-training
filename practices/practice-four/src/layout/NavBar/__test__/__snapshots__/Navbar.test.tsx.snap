// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should match the snapshot 1`] = `
<DocumentFragment>
  <nav
    class="bg-light w-full border"
  >
    <div
      class="flex flex-wrap items-center justify-between p-4"
    >
      <a
        class="text-2xl font-semibold text-gray-900"
        href="/projects"
      >
        Projects
      </a>
      <div
        class="flex items-center gap-6"
      >
        <img
          alt="Notification"
          class="cursor-not-allowed"
          src="mock-notify-icon"
        />
        <img
          alt="Help"
          class="cursor-not-allowed"
          src="mock-help-icon"
        />
        <div
          class="rounded-full cursor-not-allowed"
        >
          Mocked Avatar - HienDuong
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
